src/mmrm.cpp

```diff
@@ -60,42 +60,33 @@ Type objective_function<Type>::operator() ()
   // Convert is_spatial_int to bool.
   bool is_spatial = (is_spatial_int == 1);
   // Create the lower triangular Cholesky factor of the visit x visit covariance matrix.
-  int dim_cov_mat;
-  if (is_spatial) {
-    dim_cov_mat = 2;
-  } else {
-    dim_cov_mat = n_visits;
+  std::map<int, lower_chol_base<Type>*> chols_by_group;
+  for (int r = 0; r < n_groups; r++) {
+    // in loops using new keyword is required so that the objects stays on the heap
+    // otherwise this will be destroyed and you will get unexpected result
+    if (is_spatial) {
+      chols_by_group[r] = new lower_chol_sp_exp<Type>(theta.segment(r * theta_one_group_size, theta_one_group_size));
+    } else {
+      chols_by_group[r] = new lower_chol_nonspatial<Type>(theta.segment(r * theta_one_group_size, theta_one_group_size), n_visits, cov_type);
```

Comment:
so here we still need to differentiate between spatial and non-spatial, any chance to also unify this at this code level?

Reply:
if too complicated, not needed, or at least not in this PR

Reply:
actually, they are somehow unified. they are both class lower_chol_base, and have the same method "get_chol". So here they are kind of unified. 

If we would unify further, then we probably will make the "lower_chol_nonspatial" and "lower_chol_sp_exp" one class (then we would store the type inside, and every time depending on the covariance type choose suitable method). I did not do this at first because, I doubt for spatial covariance anything should be cached (this cache only works for the categorical time points)